codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+      else:
+        md.expiry = expiry
+        md.leavesCount = leavesCount
+        md.manifestsCids = manifestsCids
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+
+      md.some
+  )
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  leavesCount: Natural,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(treeCid, leavesCount, self.clock.now + self.defaultExpiry.seconds, manifestsCids)
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  cid: Cid,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(cid, 0, self.clock.now + self.defaultExpiry.seconds, manifestsCids)
+
+proc ensureExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  minExpiry: SecondsSince1970): Future[?!void] {.async.} =
+  ## Sets the dataset expiry to a max of two values: current expiry and `minExpiry`,
+  ## if a dataset for given `treeCid` is not currently tracked a CatchableError is thrown
+  ##
+
+  trace "Updating a dataset expiry", treeCid, minExpiry
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await modify[DatasetMetadata](self.metaDs, key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        let datasetMd = DatasetMetadata(
+          expiry: max(currDatasetMd.expiry, minExpiry),
+          leavesCount: currDatasetMd.leavesCount,
+          manifestsCids: currDatasetMd.manifestsCids,
+          checkpoint: currDatasetMd.checkpoint
+        )
+        return datasetMd.some
+      else:
+        raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " not found")
+  )
+
+
+proc recordCheckpoint*(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  # Saves progress or deletes dataset metadata if progress > leavesCount
+  #
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
 
-proc runBlockCheck(self: BlockMaintainer): Future[void] {.async.} =
-  let expirations = await self.repoStore.getBlockExpirations(
-    maxNumber = self.numberOfBlocksPerInterval,
-    offset = self.offset
+  await self.metaDs.modify(key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        if currDatasetMd.expiry != datasetMd.expiry or currDatasetMd.manifestsCids != datasetMd.manifestsCids:
+          raise newException(CatchableError, "Change in expiry detected, interrupting maintenance for dataset with treeCid " & $treeCid)
+
+        if currDatasetMd.checkpoint.progress > datasetMd.checkpoint.progress:
+          raise newException(CatchableError, "Progress should be increasing only, treeCid " & $treeCid)
+
+        if currDatasetMd.leavesCount <= datasetMd.checkpoint.progress:
```

Comment:
It will work either way with `==` and with `<=`.

If there's no additional cost to it, I prefer to have stop conditions safe, even if some bugs were introduced in the future. One future bug would be to increment `progress` always by BatchSize (even if there was less stuff deleted), in such scenario `==` wouldn't work. I know it's hypothetical so if you want I can change it.